File: UI/src/main/java/ca/warp7/rt/core/app/AppElement.java
package ca.warp7.rt.core.app;

import ca.warp7.rt.core.feature.FeatureIcon;
import ca.warp7.rt.core.feature.FeatureItemTab;
import javafx.geometry.Pos;
import javafx.scene.control.Alert;
import javafx.scene.control.ButtonType;
import javafx.scene.control.Label;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.HBox;
import org.kordamp.ikonli.javafx.FontIcon;

class AppElement {

    static HBox tabUIFromAction(AppTab tab) {

        FeatureItemTab action = tab.getFeatureItemTab();

        HBox outer = new HBox();

        HBox inner = new HBox();
        inner.setPrefWidth(20);
        inner.setAlignment(Pos.CENTER);

        FontIcon icon = new FeatureIcon(action.getIconLiteral());

        inner.getChildren().add(icon);
        tab.setIcon(icon);

        outer.setSpacing(10);
        outer.getChildren().add(inner);
        outer.getChildren().add(new Label(action.getTitle()));

        return outer;
    }

    static void showMemoryAlert() {
        String mem = String.format("Memory: %.2f MB", (Runtime.getRuntime().totalMemory()
                - Runtime.getRuntime().freeMemory()) / 1000000.0);
        Alert alert = new Alert(Alert.AlertType.INFORMATION, mem, ButtonType.OK);
        new Thread(System::gc).start();
        alert.showAndWait();
    }

    static AppTab getTeamLogo() {
        int height = 72;
        ImageView teamLogo = new ImageView();
        teamLogo.setImage(new Image("/ca/warp7/rt/res/warp7.png"));
        teamLogo.setPreserveRatio(true);
        teamLogo.setFitHeight(height);
        HBox hBox = new HBox();
        hBox.setAlignment(Pos.TOP_CENTER);
        hBox.getChildren().add(teamLogo);
        return new AppTab(hBox, height);
    }

    static String getTitle(FeatureItemTab tab) {
        String title;
        if (tab.getTabGroup() == FeatureItemTab.Group.SingleTab) title = String.format("%s", tab.getTitle());
        else {
            String id = tab.getFeatureId();
            String capId = id.substring(0, 1).toUpperCase() + id.substring(1);
            title = String.format("%s - %s", tab.getTitle(), capId);
        }
        return title;
    }
}

File: UI/src/main/kotlin/ca/warp7/rt/core/app/AppElement.kt
package ca.warp7.rt.core.app

import ca.warp7.rt.core.env.EnvUser
import ca.warp7.rt.core.env.EnvUtils
import ca.warp7.rt.core.feature.FeatureIcon
import ca.warp7.rt.core.feature.FeatureItemTab
import ca.warp7.rt.core.feature.FeatureLink
import javafx.geometry.Pos
import javafx.scene.control.Label
import javafx.scene.layout.HBox
import javafx.scene.paint.Color

internal object AppElement {

    @JvmStatic
    val teamColor: Color = Color.valueOf("1e2e4a")

    @JvmStatic
    fun tabUIFromAction(tab: AppTab): HBox {
        val action = tab.featureItemTab
        val outer = HBox()
        val inner = HBox()
        inner.prefWidth = 24.0
        inner.alignment = Pos.CENTER
        val icon = FeatureIcon(action.iconLiteral)
        icon.iconColor = teamColor
        inner.children.add(icon)
        tab.icon = icon
        outer.alignment = Pos.CENTER_LEFT
        outer.spacing = 10.0
        outer.children.add(inner)
        outer.children.add(Label(action.title))
        return outer
    }

    fun tabUIFromLink(link: FeatureLink): HBox {
        val outer = HBox()
        val inner = HBox()
        inner.prefWidth = 24.0
        inner.alignment = Pos.CENTER
        link.icon.iconColor = teamColor
        inner.children.add(link.icon)
        outer.alignment = Pos.CENTER_LEFT
        outer.spacing = 10.0
        outer.children.add(inner)
        outer.children.add(Label(link.title))
        return outer
    }


    @JvmStatic
    fun getTitle(tab: FeatureItemTab): String {
        val title: String
        if (tab.tabGroup == FeatureItemTab.Group.SingleTab)
            title = String.format("%s", tab.title)
        else {
            val id = tab.featureId
            val capId = id.substring(0, 1).toUpperCase() + id.substring(1)
            title = String.format("%s - %s", tab.title, capId)
        }
        return title
    }

    @JvmStatic
    fun getUserExplicitName(): String {
        val current = EnvUser["app.userName", EnvUtils.user]
        val name = AppUtils.getString("Setup", "Enter name (First Last):",
                current, String::isNotEmpty) ?: current
        EnvUser["app.userName"] = name
        return name
    }

    @JvmStatic
    fun getUserExplicitDevice(): String {
        val current = EnvUser["app.deviceName", EnvUtils.computerName]
        val name = AppUtils.getString("Setup", "Enter device:",
                current, String::isNotEmpty) ?: current
        EnvUser["app.deviceName"] = name
        return name
    }

    @JvmStatic
    fun updateUserAndDevice(userName: Label, deviceName: Label) {
        userName.text = when {
            !EnvUser["app.userName"] -> getUserExplicitName()
            else -> EnvUser["app.userName", "Unknown User"]
        }
        deviceName.text = when {
            !EnvUser["app.deviceName"] -> getUserExplicitDevice()
            else -> EnvUser["app.deviceName", "Unknown Device"]
        }
        EnvUser.save()
    }
}
